Issue #344: add multi-thread support to CheckstyleReportsParser

rnveach opened the pull request:
Issue #344

I feel I have been sitting on this long enough. I have not run into any issues in all the times I have used it.

romani commented:
@rnveach , please write a summary of benefit to have MT mode. Is it only for extremely large reports ? Or it is benefitial on real life examples of use that we have currently in our PRs as regression report.

rnveach commented:
The benefit of MT is that the process will now split any work that was solely done in a single thread to be done simultaneously into separate threads, allowing the execution to be split among as many available threads open to the system for however big the reports and differences get.

Parsing of the 2 XML files, base and patch as represented by O(B + P) time (basically O(2n) time), is now split into 2 different threads allowing parsing time decreased to the smallest of either O(B) and O(P), which basically amounts to now O(n) time.

Finding differences on the violations in the XML files is where the most benefit is seen. Difference finding, which is comparing 2 lists and amounts to O(B * P) time (basically an O(n^2) complexity), is now split into as small as 100 record chunks to each thread that is available on the system. This haves a dramatic effect on time proportional to how many threads are on the system, theoretically amounting to O(B * P / T) time, where T is number of threads given to the process. The more threads available and the more records to process, the more time is saved with multi-threading. I think with all the savings, this may reduce the complexity to basically O(n log n) time. ( https://stackoverflow.com/questions/1592649/examples-of-algorithms-which-has-o1-on-log-n-and-olog-n-complexities )

> Is it only for extremely large reports ?

No. Anything with over 100 violations in any XML file will see improvements. The bigger, the more improvements seen.

>  is benefitial on real life examples of use that we have currently in our PRs as regression report.

Mine was a real life example as presented in issue. The more violations a produced in the final XML, the more savings will be generated. Pitest regression ( https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-pitest-regression ) multiplies the number of violations seen by 1 check and makes the number of violations bigger than normal. Indentation also produces possibly many violations just per file and will the be the single check to show the most improvement.

Remember improvements are seen based on the more violations in the XML file. This search routine is O(n^2) complexity.

romani commented:
I did not mean general ideas of why MT mode is better.

I meant, what is reduction of execution time for CI, for example for https://github.com/checkstyle/checkstyle/pull/8913 .

rnveach commented:
> I meant, what is reduction of execution time for CI

That is explained as I go into O(n) time complexity. It the notation used to describe how many comparisons a routine does. The less the time complexity, the less comparisons done, the faster the routine generally is.

O(1) is less time than O(n)
O(n) is less time than O(n log n)
O(n log n) is less time than O(n^2)
etc...

Main issue provides some numbers an actual run but I don't have the file to reproduce and provide more details from it, I can only explain the time reduction in terms of comparisons as there are many variables.

> for example for checkstyle/checkstyle#8913 .

Can you post to exact comment because I am not seeing anything.

romani commented:
I understand notation.
But if optimize process that takes 10 seconds to be 5 seconds, in Os it awesome but in real life it not that much , especially if whole other process takes 40 minutes. 

I curious how much quicker generation of report be after merge if this update. So I provided link to PR with configs.
I need to maintainability price of code in comparison with benefits of time reduction to get report ready.

pbludov commented on patch-diff-report-tool/src/main/java/com/github/checkstyle/data/DiffReport.java:
```diff
@@ -103,16 +128,44 @@ public void addRecords(List<CheckstyleRecord> newRecords,
+                }
             }
         }
-        for (CheckstyleRecord rec2 : list2) {
```
Why this loop is removed?

rnveach replied:
This is the loop that amounts to O(N^2) complexity and is the real slowdown that needs multi-threading. The nested loop is inside `isInList`. It is now transferred to line 159 to split the list into pieces and send to the threads to complete. The rest of the loop is moved to line 270 which is inside the thread.

pbludov replied:
I don't get it at all. The code on line 159 will be executed only if the size of the list >= `SPLIT_SIZE`. Otherwise, the method will return only records that are in the list1 and are missing in the list2.

rnveach replied:
It sounds like you are missing https://github.com/checkstyle/contribution/blob/44f02903040bc7f16c834d26ed324e96b2c463a9/patch-diff-report-tool/src/main/java/com/github/checkstyle/data/DiffReport.java#L133-L134 which calls the same method with the lists swapped.

pbludov commented on patch-diff-report-tool/src/main/java/com/github/checkstyle/data/DiffReport.java:
```diff
@@ -103,16 +128,44 @@ public void addRecords(List<CheckstyleRecord> newRecords,
+            final ExecutorService executor =
+                Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+            final List<Future<List<CheckstyleRecord>>> futures = new ArrayList<>();
+            final int size = list1.size();
```
this looks as hardcoded
```java
    private static List<CheckstyleRecord> produceDiff(
            List<CheckstyleRecord> list1, List<CheckstyleRecord> list2) {
        return Stream.concat(
                StreamSupport.stream(list1.spliterator(), SPLIT_SIZE <= list1.size())
                    .filter(rec -> !isInList(list2, rec)),
                StreamSupport.stream(list2.spliterator(), SPLIT_SIZE <= list2.size())
                    .filter(rec -> !isInList(list1, rec))
            )
            .collect(Collectors.toList());
    }
```

rnveach replied:
Is this multi-threaded? The whole issue is we need this multi-threaded to reduce processing time looping through the 2 lists.

pbludov replied:
` StreamSupport.stream(list1.spliterator(), SPLIT_SIZE <= list1.size())` is basically `list1.stream()` if the size is less then `SPLIT_SIZE`. Otherwise it is `list1.parallelStream()`

rnveach replied:
I am not familar with streams but I assume parallel is multi-threaded. I will compare and see how they stack against each other.

pbludov commented:
@rnveach what about such optimization:
```java
        List<CheckstyleRecord> shortest;
        List<CheckstyleRecord> longest;
        if (list1.size() < list2.size()) {
            shortest = list1;
            longest = list2;
        }
        else {
            shortest = list2;
            longest = list1;
        }

        // O(log(N))
        Set<CheckstyleRecord> recordSet =
                StreamSupport.stream(shortest.spliterator(), SPLIT_SIZE <= shortest.size()) // Do we need threading here?
            .collect(Collectors.toCollection(ConcurrentSkipListSet::new));

        // O(M * log(N)), M > N
        List<CheckstyleRecord> diff =
                StreamSupport.stream(longest.spliterator(), SPLIT_SIZE <= longest.size())
            .filter(rec -> !recordSet.remove(rec)) // Remove from both collections if matches
            .collect(Collectors.toCollection(ArrayList::new));

        diff.addAll(recordSet);
        return diff;
```
Could you run your benchmark on this code?

rnveach commented:
> what about such optimization:

I can add it to the list to benchmark but I don't understand the code to know what it is doing. Is this a replacement for `produceDiffEx` or `produceDiff`?

pbludov commented:
> Imagine both lists are of infinite size, it doesn't matter who is smaller. This is how I read `N` in this complexities since it isn't a direct number. It only matters what the final complexity is and if it can be reduced below `O(N^2)` or `O(N log N)`.

In this case, may be we should keep these lists *sorted*? It is easy to compare two sorted lists. Or even store them in a sorted collection, like
```java
    private Map<String, ConcurrentSkipListSet<CheckstyleRecord>> records
```
?

rnveach commented:
> In this case, may be we should keep these lists sorted?

It looks like we sorted the results after the diff. https://github.com/checkstyle/contribution/blob/44f02903040bc7f16c834d26ed324e96b2c463a9/patch-diff-report-tool/src/main/java/com/github/checkstyle/data/DiffReport.java#L113

If we are going to sort beforehand, wouldn't it be better to sort the records as they are added to the list?

romani commented:
@rnveach , please share comparison of executions, you might use Indentation Check to make sure that reports are huge in diff.

pbludov commented:
Please take a look to my PR #521
